benefits/enrollment/views.py

```diff
@@ -84,25 +84,56 @@ def index(request):
+        group_id = eligibility.group_id
 
         try:
-            client.link_concession_group_funding_source(funding_source_id=funding_source.id, group_id=eligibility.group_id)
-        except HTTPError as e:
-            # 409 means that customer already belongs to a concession group.
-            # the response JSON will look like:
-            # {"errors":[{"detail":"Conflict (409) - Customer already belongs to a concession group."}]}
-            if e.response.status_code == 409:
+            group_funding_sources = client.get_concession_group_linked_funding_sources(group_id)
+            matching_group_funding_source = None
+            for group_funding_source in group_funding_sources:
+                if group_funding_source.id == funding_source.id:
+                    matching_group_funding_source = group_funding_source
+                    break
+
+            # funding source is already linked to the group
+            if matching_group_funding_source:
+                # if no expiration date, return success
+                if matching_group_funding_source.concession_expiry is None:
+                    analytics.returned_success(request, group_id)
+                    return success(request)
+                # else if expiration date has passed, then re-enroll
+                elif expiration.is_expired(matching_group_funding_source):
+                    client.link_concession_group_funding_source(funding_source_id=funding_source.id, group_id=group_id)
+                    analytics.returned_success(request, group_id)
+                    return success(request)
+                # else expiration date hasn't passed, so calculate re-enrollment date and show user
+                else:
+                    reenrollment_date = expiration.calculate_reenrollment_date(  # noqa
+                        matching_group_funding_source, eligibility.expiration_reenrollment_days
+                    )
+
+                    analytics.returned_error(request, "Funding source already enrolled and has not expired")
+                    # todo for #1921: show reenrollment_date to user
+            # funding source has not been linked to the group yet
+            else:
+                # if eligibility_type does not supports_expiration, then just link it
+                if not eligibility.supports_expiration:
+                    client.link_concession_group_funding_source(funding_source_id=funding_source.id, group_id=group_id)
+                # else eligibility_type supports_expiration, so calculate expiration date from today and include in request
+                else:
+                    expiry_date = expiration.calculate_expiry_date(matching_group_funding_source, eligibility.expiration_days)
+                    client.link_concession_group_funding_source(
+                        funding_source_id=funding_source.id, group_id=group_id, expiry_date=expiry_date
+                    )
```

Comment:
@thekaveman I put all this business logic right here in the view function 1) so I could get it written down for starters, and 2) so that we could see what might make sense to extract out as helper functions.

I actually think this looks ok. It's kind of long, but in terms of level of abstraction, it seems appropriate to have the API calls here and delegate out the date calculation to a separate module.

Thoughts? Also are the comments helpful or just noisy?

Reply:
By "all this business logic" I mean lines [90-120](https://github.com/cal-itp/benefits/pull/1973/files#diff-b67985b5acabad774a30a1fccbb912e611d4056d38316cd6614ce601f6930335R90-R120)